src/cc/libbpf.c

```diff
@@ -133,6 +133,42 @@ int bpf_delete_elem(int fd, void *key)
   return syscall(__NR_bpf, BPF_MAP_DELETE_ELEM, &attr, sizeof(attr));
 }
 
+int bpf_get_first_key(int fd, void *key, size_t key_size, size_t value_size)
+{
+  union bpf_attr attr;
+  void *value;
+  int i, res;
+
+  memset(&attr, 0, sizeof(attr));
+  attr.map_fd = fd;
+  attr.key = 0;
+  attr.next_key = ptr_to_u64(key);
+
+  // 4.12 and above kernel supports passing NULL to BPF_MAP_GET_NEXT_KEY
+  // to get first key of the map. For older kernels, the call will fail.
+  res = syscall(__NR_bpf, BPF_MAP_GET_NEXT_KEY, &attr, sizeof(attr));
+  if (res < 0 && errno == EFAULT) {
+    // Fall back to try to find a non-existing key.
+    static unsigned char try_values[3] = {0, 0xff, 0x55};
+    value = malloc(value_size);
+    for (i = 0; i < 3; i++) {
+      memset(key, try_values[i], key_size);
+      if (bpf_lookup_elem(fd, key, value) == 0)
+        continue;
+      memset(&attr, 0, sizeof(attr));
```

Comment:
@4ast Can I safely assume the first syscall didn't modify `attr` (currently yes) so that I don't need to re-initialize it here?